### PR TITLE
ReadOnly option for Invoke-DbaQuery

### PR DESCRIPTION
**Pull request by michalporeba**
## Type of Change

New Feature #3451
 

### Purpose
 
To be able to easily run queries read only queries on secondary nodes in AG

### Approach

When `Invoke-DbaQuery` is invoked with `-ReadOnly` parameter the ReadOnly application intent is set which allows the read only queries to succeed even if the specific database on the current instance is a read only replica and not primary. 

### Commands to test

`Invoke-DbaQuery ... -ReadOnly`

### CODE REVIEW

**Comment by wsmelton** [comment]
Use of custom properties for the connection string should be done via `Connect-DbaInstance … -ApplicationIntent ReadOnly` and then passing in that server object to the command.

The internal function should not be touched to support this feature.

**Comment by wsmelton** [comment]
Hit enter a bit to early....I see no reason to modify an internal function to support this feature/ability as this function (1) is used with everything and (2) we don't need to update an internal function in order to support one feature that only one command will ever use.

**Comment by wsmelton** [comment]
Which this was all discussed in that issue as well.

**Comment by wsmelton** [comment]
Yeah, also testing (at least on PS Core) this change has broken any command I use with a SQL Login. Not all of our test utilize a SQL Login to test the credential so assuming the checks are covering every scenario would be false.

image

Extensive testing would have to be done if I'm overruled and this feature is approved.

**Comment by michalporeba** [comment]
Thanks Shawn for your comments. I see one test is failing and I will look more into it this afternoon. I see the internal function is light on the tests, but if it is used so much it will be tested with every other test of the function that is using it. Still, it would be good to improve testing I think. 

As for the discussion about the feature itself it would be good to go to the original issue #3451.

File: functions/Invoke-DbaQuery.ps1 [hunk 1]
```diff
@@ -118,7 +121,8 @@ function Invoke-DbaQuery {
         [switch]$MessagesToOutput,
         [parameter(ValueFromPipeline)]
         [Microsoft.SqlServer.Management.Smo.Database[]]$InputObject,
-        [switch]$EnableException
+        [switch]$EnableException,
+        [switch]$ReadOnly
```
Comment (wsmelton) [inline functions/Invoke-DbaQuery.ps1]: `EnableException` should always be the last parameter, just for tab purposes.

File: tests/Invoke-DbaQuery.Tests.ps1 [hunk 2]
```diff
@@ -4,10 +4,10 @@ Write-Host -Object "Running $PSCommandPath" -ForegroundColor Cyan
 
 Describe "$CommandName Unit Tests" -Tag 'UnitTests' {
     Context "Validate parameters" {
-        $paramCount = 13
+        $paramCount = 14
```
Comment (wsmelton) [inline tests/Invoke-DbaQuery.Tests.ps1]: Check #4695 and you might go ahead and adjust this test to match the new pattern for unit testing.

File: functions/Invoke-DbaQuery.ps1 (hunk 1 above)
Comment (michalporeba) [reply functions/Invoke-DbaQuery.ps1]: done, I'll be pushing the change with the few more updates shortly

File: tests/Invoke-DbaQuery.Tests.ps1 (hunk 2 above)
Comment (michalporeba) [reply tests/Invoke-DbaQuery.Tests.ps1]: I'll look into it.

Comment (michalporeba) [reply tests/Invoke-DbaQuery.Tests.ps1]: Done.

**Comment by michalporeba** [comment]
I'm really out of shape. So many commits for so little. Sorry :)

**Comment by potatoqualitee** [comment]
i do the same @michalporeba 🤗

**Comment by potatoqualitee** [comment]
squash and merge, now all 1 💯

**Comment by wsmelton** [comment]
Little commits are best and user more commonly than you may think. It makes it easier to revert or remove those commits around breaking changes if they are discovered during testing.

**Pull request merged**